trace-forward/src/Trace/Forward/Network/Acceptor.hs

```diff
@@ -0,0 +1,200 @@
+  InitiatorProtocolOnly $
+    MuxPeerRaw $ \channel ->
+      runPeer
+        (acceptorTracer config)
+        (Acceptor.codecTraceForward CBOR.encode CBOR.decode
+                                    CBOR.encode CBOR.decode
+                                    CBOR.encode CBOR.decode)
+        channel
+        (Acceptor.traceAcceptorPeer $
+          acceptorActions config loHandler niHandler True False)
+
+acceptorActions
+  :: (CBOR.Serialise lo,
+      ShowProxy lo,
+      Typeable lo)
+  => AcceptorConfiguration lo -- ^ Acceptor's configuration.
+  -> ([lo] -> IO ())          -- ^ The handler for accepted 'TraceObject's.
+  -> (NodeInfo -> IO ())      -- ^ The handler for accepted info about the node.
+  -> Bool                     -- ^ The flag for node's info request: only once in the beginning.
+  -> Bool                     -- ^ The flag for end of session: if 'True', 'SendMsgDone' will be sent.
+  -> Acceptor.TraceAcceptor lo IO ()
+acceptorActions config@AcceptorConfiguration{whatToRequest, shouldWeStop} loHandler niHandler askForNI False =
+  -- We are able to send request for:
+  -- 1. node's info,
+  -- 2. new 'TraceObject's.
+  -- But request for node's info should be sent only once (in the beginning of session).
+  if askForNI
+    then
+      Acceptor.SendMsgNodeInfoRequest $ \replyWithNI -> do
+        runActionWithTimer (niHandler replyWithNI) limitInSecs
+        readIORef shouldWeStop <&> acceptorActions config loHandler niHandler False
+    else
+      Acceptor.SendMsgTraceObjectsRequest TokBlocking whatToRequest $ \replyWithTraceObjects -> do
+        runActionWithTimer (loHandler $ getTraceObjects replyWithTraceObjects) limitInSecs
+        readIORef shouldWeStop <&> acceptorActions config loHandler niHandler False
+ where
+  limitInSecs = 15
+
+acceptorActions _ _ _ _ True =
+  Acceptor.SendMsgDone $ return ()
+
+-- | The configuration provides two external actions: 'loHandler' and 'niHandler'.
+--   Since we cannot make any assumptions about them, it is possible that some of
+--   them will take a too long time to perform. But in this case, the protocol may
+--   "freeze" and won't be able to check 'shouldWeStop' flag soon.
+--
+--   To prevent this, we run a stop timer in parallel with action, so:
+--
+--   1. If this action ends first (the normal scenario), its thread finishes first
+--      and stop timer's thread will be canceled.
+--   2. If the stop timer ends first, it means that action is still running
+--      (something is wrong with it), so stop timer's thread finishes and action's
+--      thread will be canceled.
+runActionWithTimer
+  :: IO () -- ^ An action to be performed.
+  -> Int   -- ^ Number of seconds for stop timer.
+  -> IO ()
+runActionWithTimer action timerInSecs = do
+  flag <- registerDelay delayInMicroSecs
+  race_ action $ timer flag
+ where
+  delayInMicroSecs = timerInSecs * 1000000
+
+  timer flag' =
+    unlessM (readTVarIO flag') $ do
+      sleep 0.01
+      timer flag'
```

Comment:
```suggestion
runActionWithTimer action timerInSecs = do
    v <- registerDelay (delayInSecs * 1_000_000)
    withAsync action $ \a -> atomically $
       waitSTM a `orElse` readTVar v >>= check
```

Reply:
Thank you a lot for the suggestions @coot !

Reply:
IMHO, it's much harder to read/understand...

Reply:
Good question. First of all it a declarative description what it does rather than a procedural. So I'd argue that it is easier to understand, although that might depend how familiar one is with stm. One could also wrap it with [`FirstToFinish`](https://input-output-hk.github.io/ouroboros-network/monoidal-synchronisation/Data-Monoid-Synchronisation.html#t:FirstToFinish) and use monoid instance insead  of `orElse`; or use `<|>` intead of `orElse`.

Beside it being declarative, its also more efficient. There's no semi-busy loop which needs to be handled by cpu every 0.1sec. The stm transaction will wake up only once either if the thread is done or when the timeout expires.